### PR TITLE
Added mpi backend plugin

### PR DESCRIPTION
**Pull request by evalsocket**
# TL;DR
- Added MPI ackend plugin 
-
## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -46,9 +47,34 @@ require (
...
-	k8s.io/client-go v0.20.2
+	k8s.io/client-go v10.0.0+incompatible
 	k8s.io/utils v0.0.0-20210111153108-fddb29f9d009
 	sigs.k8s.io/controller-runtime v0.8.2
 )
 
-replace github.com/aws/amazon-sagemaker-operator-for-k8s => github.com/aws/amazon-sagemaker-operator-for-k8s v1.0.1-0.20210303003444-0fb33b1fd49d
+replace (
+	github.com/aws/amazon-sagemaker-operator-for-k8s => github.com/aws/amazon-sagemaker-operator-for-k8s v1.0.1-0.20210303003444-0fb33b1fd49d
+	k8s.io/api => k8s.io/api v0.20.2
```
Comment (kumare3) [inline go.mod:57]: this looks bad!!!!!!!!!!!!

Comment (evalsocket) [reply go.mod:57]: yeah, I am aware of the issue.

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go [hunk 2]
```diff
@@ -120,6 +157,47 @@ func GetLogs(taskType string, name string, namespace string,
...
+		})
+		if err != nil {
+			return nil, err
+		}
+		taskLogs = append(taskLogs, workerLog.TaskLogs...)
+	}
+	// get all worker servers logs
+	for psReplicaIndex := int32(0); psReplicaIndex < launcherReplicasCount; psReplicaIndex++ {
+		// TODO: Add launcher pod name, {POD_NAME}-launcher-HASH
+		psReplicaLog, err := logPlugin.GetTaskLogs(tasklog.Input{
```
Comment (evalsocket) [inline go/tasks/plugins/k8s/kfoperators/common/common_operator.go:188]: @EngHabu @kumare3 I have one question. MPI operator creates a pod with a name template like `{POD_NAME}-launcher-HASH` and when we create a log we need the exact name of the pod. Do you have any suggestion to resolve it.

File: go.mod (hunk 1 above)
Comment (evalsocket) [reply go.mod:58]: @kumare3  https://github.com/kubernetes/kubernetes/issues/79384

Comment (kumare3) [reply go.mod:58]: i dont think we need kube api

Comment (EngHabu) [reply go.mod:58]: Just found out there is a `/v2` version of the package that Yuvraj can use...
https://github.com/kubeflow/mpi-operator/tree/master/v2

`go get github.com/kubeflow/mpi-operator/v2`

It's currently on 0.19... we can probably fork it and propose a PR to upgrade to 0.20

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go [hunk 3]
```diff
@@ -11,15 +11,33 @@ import (
...
 const (
 	TensorflowTaskType = "tensorflow"
+	MPITaskType        = "mpi"
 	PytorchTaskType    = "pytorch"
 )
 
+func ExtractMPICurrentCondition(jobConditions []mpiOp.JobCondition) (mpiOp.JobCondition, error) {
+	if jobConditions != nil {
+		sort.Slice(jobConditions, func(i, j int) bool {
+			return jobConditions[i].LastTransitionTime.Time.After(jobConditions[j].LastTransitionTime.Time)
```
Comment (katrogan) [inline go/tasks/plugins/k8s/kfoperators/common/common_operator.go:28]: can `LastTransitionTime` or `LastTransitionTime.Time` ever be nil?

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go [hunk 4]
```diff
@@ -11,15 +11,33 @@ import (
...
+		})
+
+		for _, jc := range jobConditions {
```
Comment (katrogan) [inline go/tasks/plugins/k8s/kfoperators/common/common_operator.go:31]: why sort above if you're just going to iterate through jobConditions anyways?

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go [hunk 5]
```diff
@@ -55,6 +73,25 @@ func GetPhaseInfo(currentCondition commonOp.JobCondition, occurredAt time.Time,
...
+	switch currentCondition.Type {
+	case mpiOp.JobCreated:
+		return pluginsCore.PhaseInfoQueued(occurredAt, pluginsCore.DefaultPhaseVersion, "JobCreated"), nil
+	case mpiOp.JobRunning:
+		return pluginsCore.PhaseInfoRunning(pluginsCore.DefaultPhaseVersion, &taskPhaseInfo), nil
+	case mpiOp.JobSucceeded:
+		return pluginsCore.PhaseInfoSuccess(&taskPhaseInfo), nil
+	case mpiOp.JobFailed:
+		details := fmt.Sprintf("Job failed:\n\t%v - %v", currentCondition.Reason, currentCondition.Message)
+		return pluginsCore.PhaseInfoRetryableFailure(flyteerr.DownstreamSystemError, details, &taskPhaseInfo), nil
```
Comment (katrogan) [inline go/tasks/plugins/k8s/kfoperators/common/common_operator.go:87]: is an mpi job failure always retriable?

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go [hunk 6]
```diff
@@ -94,27 +131,30 @@ func GetLogs(taskType string, name string, namespace string,
...
-	// get all parameter servers logs
-	for psReplicaIndex := int32(0); psReplicaIndex < psReplicasCount; psReplicaIndex++ {
-		psReplicaLog, err := logPlugin.GetTaskLogs(tasklog.Input{
-			PodName:   name + fmt.Sprintf("-psReplica-%d", psReplicaIndex),
-			Namespace: namespace,
-		})
-		if err != nil {
-			return nil, err
+
+	if taskType != MPITaskType && taskType != PytorchTaskType {
```
Comment (katrogan) [inline go/tasks/plugins/k8s/kfoperators/common/common_operator.go:135]:
```suggestion
	if taskType == MPITaskType || taskType == PytorchTaskType {
            return taskLogs, nil
      }
```

so you don't have to indent the whole block below

File: go/tasks/plugins/k8s/kfoperators/mpi/mpi.go [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+	err = utils.UnmarshalStruct(taskTemplate.GetCustom(), &mpiTaskExtraArgs)
+	if err != nil {
+		return nil, flyteerr.Errorf(flyteerr.BadTaskSpecification, "invalid TaskSpecification [%v], Err: [%v]", taskTemplate.GetCustom(), err.Error())
+	}
+
+	podSpec, err := flytek8s.ToK8sPodSpec(ctx, taskCtx)
+	if err != nil {
+		return nil, flyteerr.Errorf(flyteerr.BadTaskSpecification, "Unable to create pod spec: [%v]", err.Error())
+	}
+	workers := mpiTaskExtraArgs.GetNumWorkers()
```
Comment (katrogan) [inline go/tasks/plugins/k8s/kfoperators/mpi/mpi.go:64]: nit, move lines 64-66 to just below 58 so that everything is organized together

File: go/tasks/plugins/k8s/kfoperators/mpi/mpi.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+			Kind:       mpi.Kind,
+			APIVersion: mpi.SchemeGroupVersion.String(),
+		},
+		Spec: jobSpec,
+	}
+
+	return job, nil
+}
+
+// Analyses the k8s resource and reports the status as TaskPhase. This call is expected to be relatively fast,
```
Comment (katrogan) [inline go/tasks/plugins/k8s/kfoperators/mpi/mpi.go:105]:
```suggestion
// Analyzes the k8s resource and reports the status as TaskPhase. This call is expected to be relatively fast,
```

File: go/tasks/plugins/k8s/kfoperators/mpi/mpi.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+// any operations that might take a long time (limits are configured system-wide) should be offloaded to the
+// background.
+func (mpiOperatorResourceHandler) GetTaskPhase(_ context.Context, pluginContext k8s.PluginContext, resource client.Object) (pluginsCore.PhaseInfo, error) {
+	app := resource.(*mpi.MPIJob)
```
Comment (katrogan) [inline go/tasks/plugins/k8s/kfoperators/mpi/mpi.go:109]: should you check whether the cast returns ok too and error if not?

File: go/tasks/plugins/k8s/kfoperators/mpi/mpi.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+	numWorkers := app.Spec.MPIReplicaSpecs[mpi.MPIReplicaTypeWorker].Replicas
+	numLauncherReplicas := app.Spec.MPIReplicaSpecs[mpi.MPIReplicaTypeLauncher].Replicas
+
+	taskLogs, err := common.GetLogs(common.MPITaskType, app.Name, app.Namespace,
+		*numWorkers, *numLauncherReplicas, 0)
```
Comment (katrogan) [inline go/tasks/plugins/k8s/kfoperators/mpi/mpi.go:115]: do you need to defensively check whether `numWorkers` or `numLauncherReplicas` might be nil before dereferencing?

File: go/tasks/plugins/k8s/kfoperators/common/common_operator.go [hunk 11]
```diff
@@ -11,15 +11,33 @@ import (
...
 	v1 "k8s.io/api/core/v1"
 )
 
 const (
 	TensorflowTaskType = "tensorflow"
+	MPITaskType        = "mpi"
 	PytorchTaskType    = "pytorch"
 )
 
+func ExtractMPICurrentCondition(jobConditions []mpiOp.JobCondition) (mpiOp.JobCondition, error) {
```
Comment (katrogan) [inline go/tasks/plugins/k8s/kfoperators/common/common_operator.go:25]: gah this is annoying you have to duplicate all these tasks for mpi job conditions
